ARCs/arc-0058.md

```diff
@@ -0,0 +1,410 @@
+  }
+
+  /**
+   * Verify the abstracted account is rekeyed to this app
+   */
+  verifyAuthAddr(): void {
+    assert(this.address.value.authAddr === this.getAuthAddr());
+  }
+
+  /**
+   * Rekey the abstracted account to another address. Primarily useful for rekeying to an EOA.
+   *
+   * @param addr The address to rekey to
+   * @param flash Whether or not this should be a flash rekey. If true, the rekey back to the app address must done in the same txn group as this call
+   */
+  rekeyTo(addr: Address, flash: boolean): void {
+    verifyAppCallTxn(this.txn, { sender: this.admin.value });
+
+    sendPayment({
+      sender: this.address.value,
+      receiver: addr,
+      rekeyTo: addr,
+      note: 'rekeying abstracted account',
+    });
+
+    if (flash) this.verifyRekeyToAbstractedAccount();
+  }
+
+  /**
+   * Temporarily rekey to an approved plugin app address
+   *
+   * @param plugin The app to rekey to
+   */
+  rekeyToPlugin(plugin: Application): void {
+    assert(this.plugins(plugin).exists);
+
+    sendPayment({
+      sender: this.address.value,
+      receiver: this.address.value,
+      rekeyTo: plugin.address,
+      note: 'rekeying to plugin app',
+    });
+
+    this.verifyRekeyToAbstractedAccount();
+  }
+
+  /**
+   * Temporarily rekey to a named plugin app address
+   *
+   * @param name The name of the plugin to rekey to
+   */
+  rekeyToNamedPlugin(name: string): void {
+    this.rekeyToPlugin(this.namedPlugins(name).value);
+  }
+
+  /**
+   * Change the admin for this app
+   *
+   * @param newAdmin The new admin
+   */
+  changeAdmin(newAdmin: Account): void {
+    verifyTxn(this.txn, { sender: this.admin.value });
+    this.admin.value = newAdmin;
+  }
+
+  /**
+   * Add an app to the list of approved plugins
+   *
+   * @param app The app to add
+   */
+  addPlugin(app: Application): void {
+    verifyTxn(this.txn, { sender: this.admin.value });
+
+    this.plugins(app).create(0);
+  }
+
+  /**
+   * Remove an app from the list of approved plugins
+   *
+   * @param app The app to remove
+   */
+  removePlugin(app: Application): void {
+    verifyTxn(this.txn, { sender: this.admin.value });
+
+    this.plugins(app).delete();
+  }
+
+  /**
+   * Add a named plugin
+   *
+   * @param app The plugin app
+   * @param name The plugin name
+   */
+  addNamedPlugin(app: Application, name: string): void {
+    verifyTxn(this.txn, { sender: this.admin.value });
+
+    assert(!this.namedPlugins(name).exists);
+    this.namedPlugins(name).value = app;
+    this.plugins(app).create(0);
+  }
+
+  /**
+   * Remove a named plugin
+   *
+   * @param name The plugin name
+   */
+  removeNamedPlugin(name: string): void {
+    verifyTxn(this.txn, { sender: this.admin.value });
+
+    const app = this.namedPlugins(name).value;
+    this.namedPlugins(name).delete();
+    this.plugins(app).delete();
+  }
+}
+```
+https://github.com/joe-p/account_abstraction.git
+
+TODO: Migrate to ARC repo, but waiting until development has settled.
```

Comment:
Eventually I'd like to link to the TEAL in the ARC but don't want to merge it into this repo yet because it'll be a PITA to continuously update it as there are changes being made, thus the TODO and link to the github